### ISSUE TITLE
[CODE] fault-tolerant inference (client side)

### ISSUE THREAD

**Issue by justheuristic** [bigscience-workshop/petals#7]
__Why:__ this is the main "value added" by LM api: users can open colab notebook and run a gigantic model.

Depends on learning-at-home/bloom-demo#3 

Solution (sketch)
Each client holds embeddings/logits on their side (loads a special a huggingface model)

Inputs: prefix(tokens), sampling parameters (top-k, top-p, etc)

Init:
1. tokenize prefix tokens and compute embeddings
2. for each stage, send out K initial requests to random servers in parallel. (no tensors are sent yet). Servers respond with their throughput sizes, queue sizes, latency. Pick the best server in each group.
   - figure out how servers represent themselves in DHT
   - [heuristic] if the chosen server has several consecutive stages, always choose to run all these stages.
3. client finds a sequence of servers that collectively have all model layers
   - [optional] record non-selected servers locally as backups for generation
5. client runs prefix through all stages, __stores inter-stage activations__

Generation:
6. get last prefix token embedding from final stage
7. run through logits, run top-k/top-p/whatever inference
   - optional: use knn (e.g. hnsw from faiss) to quickly find nearest token; alternative = colab gpu
8. embed this token using local embeddings
9. send it through one pipeline stage at a time
   - on each stage, add one more vector to prefix_embeddings
10. GOTO 6

Fault recovery:
- If a given pipeline stage fails, find a replacement stage using the same procedure as during initial stage selection
- feed the new stage with __locally_saved__ intermediate embeddings that serve as inputs to that stage
- continue inferencing normally

**Comment by justheuristic** [comment]
Implemented by @borzunov  in #91
